text/trace/0219-payload-collection.md

```diff
@@ -0,0 +1,341 @@
+
+The API & functionality of current Span attributes will remain the same, as they will
+still be used for collecting general-purpose, non-payload data. Therefore, the proposed changes are **non-breaking**. The only potential breaking change is regarding certain semantic conventions which may fit better as payload attributes, such as `db.statement`.
+
+### OTLP Updates
+
+We describe a prototype of additions to OTLP, to support encoding JSON-like objects in spans, together with some
+extra metadata regarding the original plain payload. This prototype is likely to be changed during specifications and formats discussions
+but hopefully could set basic characteristics.
+
+We propose using a similar structure to the native protobuf [Struct](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/struct.proto#L51) message,
+(which is a general representation of a JSON object), with some embedded metadata -
+
+```protobuf
+message Value {
+    // The kind of value.
+    oneof kind {
+      // Represents a null value.
+      NullValue null_value = 1;
+      // Represents a double value.
+      double number_value = 2;
+      // Represents a string value.
+      string string_value = 3;
+      // Represents a boolean value.
+      bool bool_value = 4;
+      // Represents a structured value.
+      MapValue map_value = 5;
+      // Represents a repeated `Value`.
+      ListValue list_value = 6;
+    }
+
+    // Set only if the original value is shortened, for supported types -
+    // > string: original number of characters
+    // > ListValue: original number of items
+    // > MapValue: original number of keys
+    int64 original_length = 7;
+
+    // Set only for MapValue, in case some of the keys were dropped
+    repeated string dropped_keys = 8; 
+}
+
+message MapValue {
+    // Note - we can consider using a repeated key-value for performance
+    map<string, Value> fields = 1;
+}
+
+message ListValue {
+    repeated Value values = 1;
+}
+```
+
+Then, we define a payload attribute which also includes extra metadata regarding the encoding -
+
+```protobuf
+message PayloadAttribute {
+    string key = 1;
+    Value value = 2;
+
+    // Optional - the original bytes encoding type of this value (e.g. json/yaml/avro/csv)
+    string original_encoding = 2;
+    // Optional - the size of the value as bytes encoded (including dropped data)
+    int64 encoded_size = 3;
+}
+```
+
+And the payloads attributes are added to the Span message as -
+
+```protobuf
+repeated PayloadAttributes payload_attributes = ...
+```
+
+### API Example
+
+Now let's see how we can define and use an API to set payload attributes
+(this example uses Python):
+
+```python
+# Added method to `Span` class
+def add_payload_attribute(
```

Comment:
an alternative would be to have a special SDK-provided wrapper type passed to regular set_attribute, e.g.
```
p := otel.Payload(...)
span.set_attribute(key, p)
```

In this form you have better extensibility because new fields could be added to Payload type in the future, which you cannot do with function arguments without overloading.

Reply:
I think it's a great idea to explore